### ISSUE TITLE
Suggestion: Role IDs should not be hardcoded in constants.

### ISSUE THREAD

**Issue by michael-milette** [michael-milette/moodle-filter_filtercodes#31]
I suggest that you get the roles out the database based on their archetype. You may find more than one role with the archetype student, or editingteacher. If not, you could simply search them based on their shortnames: student, editingteacher, rather than their supposed IDs. The IDs will generally be the ones you have set, the same way the admin user ID is usually 2, and guest 1, but that's an assumption that should not be made.

From issue #9 by @FMCorz

**Comment by michael-milette** [comment]
This suggestion has been implemented in #30 .

**Comment by FMCorz** [comment]
Hi Michael, well done. Alternatively you could use `archetype`, they can be useful for an admin when they would like to set several `student` roles for different purpose, they would then all be of `student` archetype.

https://github.com/moodle/moodle/blob/master/lib/db/install.xml#L982

The default values for capabilities in `db/access.php` also refer to the archetypes.

**Comment by michael-milette** [comment]
Hi @FMCorz ,

I actually started off by using archetypes however I soon realized that multiple roles can have the same archetype. When you look up a user's roles, they are referring to role ID's which refer to a unique shortname, not a unique archetype.

After writing the above explanation, I am starting to realize the value of what you are suggesting. Though shortnames can change, the archetypes are pretty much set and would not as easily be affected by modifications to roles.

I've started re-writing that part of the code to use archetypes. Seems to work but I still need to fix the case where the user switched roles. I will let you know as soon as I am done.

Thanks Fred, I really appreciate your feedback.
